firedrake/supermeshing.py

```diff
@@ -0,0 +1,432 @@
+# Code for projections and other fun stuff involving supermeshes.
+import firedrake
+import ctypes
+import os
+from firedrake.supermeshimpl import assemble_mixed_mass_matrix as ammm
+from firedrake.mg.utils import get_level
+from firedrake.petsc import PETSc
+from firedrake.mg.kernels import to_reference_coordinates, compile_element
+from firedrake.utility_meshes import UnitTriangleMesh, UnitTetrahedronMesh
+from firedrake.functionspace import FunctionSpace
+from firedrake.assemble import assemble
+from firedrake.ufl_expr import TestFunction, TrialFunction
+import firedrake.mg.utils as utils
+import ufl
+from ufl import inner, dx
+import numpy
+from pyop2.datatypes import ScalarType
+from pyop2.sparsity import get_preallocation
+from pyop2.compilation import load
+from pyop2.mpi import COMM_SELF
+from pyop2.utils import get_petsc_dir
+
+
+__all__ = ["assemble_mixed_mass_matrix"]
+
+
+class BlockMatrix(object):
+    def __init__(self, mat, dimension):
+        self.mat = mat
+        self.dimension = dimension
+
+    def mult(self, mat, x, y):
+        sizes = self.mat.getSizes()
+
+        for i in range(self.dimension):
+            start = i
+            stride = self.dimension
+
+            xa = x.array_r[start::stride]
+            ya = y.array_r[start::stride]
+            xi = PETSc.Vec().createWithArray(xa, size=sizes[1], comm=x.comm)
+            yi = PETSc.Vec().createWithArray(ya, size=sizes[0], comm=y.comm)
+            self.mat.mult(xi, yi)
+            y.array[start::stride] = yi.array_r
+
+
+def assemble_mixed_mass_matrix(V_A, V_B):
+    """
+    Construct the mixed mass matrix of two function spaces,
+    using the TrialFunction from V_A and the TestFunction
+    from V_B.
+    """
+
+    if len(V_A) > 1 or len(V_B) > 1:
+        raise NotImplementedError("Sorry, only implemented for non-mixed spaces")
+
+    if V_A.ufl_element().mapping() != "identity" or V_B.ufl_element().mapping() != "identity":
+        msg = """
+Sorry, only implemented for affine maps for now. To do non-affine, we'd need to
+import much more of the assembly engine of UFL/TSFC/etc to do the assembly on
+each supermesh cell.
```

Comment:
Argues for composition of tsfc kernels in a better way, plus dual evaluation in finat, etc...